Fixed #27991 -- Add obj arg to InlineModelAdmin.has_add_permission().

jdufresne opened the pull request:
Thanks Vladimir Ivanov for the initial patch.

https://code.djangoproject.com/ticket/27991

timgraham commented:
Generally the system check framework isn't used for deprecation warnings. I see the advantage that warnings are shown outside of the request/response cycle but my concern is that the method for silencing the warnings is different. If a third-party app has this issue, for example, then all projects that use that app have to modify their settings to silence the warning. Maybe it's best to use the check framework and use `warnings.warn()` instead of creating a `Warning`.

jdufresne commented:
I made the suggested change to use `warnings.warn()`. That works fine for me.

I think if usage errors like this can be caught during static analysis, it is usually easier for the programmer to find them all. As a separate enhancement, what would you think about adding a new check message `Deprecated`, `Deprecation`, or similar for these situations? If you agree, I could open a ticket to propose this enhancement.

timgraham commented:
I don't see the advantage of adding a new check message.

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2369,6 +2372,17 @@ The ``InlineModelAdmin`` class adds:
+   Should return ``True`` if adding a child object inline is permitted,
+   ``False`` otherwise. ``obj`` is the parent object being edited. If ``obj``
+   is ``None``, should return ``True`` or ``False`` when adding a new parent
+   object.
```
I don't understand the _If `obj` is `None`_ case. I think it needs expanding slightly. 

@icu0755 was right on the original PR: beyond the addition here, there are no docs for the `InlineModelAdmin` versions of these methods. Whilst the signature is the same, the semantics are different. Perhaps that is a different ticket, or commit at least (to add  `has_change_permission` and `has_delete_permission`).

jdufresne replied:
> I don't understand the If obj is None case. I think it needs expanding slightly.

I've updated this to the following. Let me know if you think it is an improvement. I'm open to additional wording suggestions.

> Should return `True` if adding a child object inline is permitted, `False` otherwise. `obj` is the parent object being edited. When adding a new parent object, `obj` is `None`.

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1887,11 +1887,15 @@ templates used by the :class:`ModelAdmin` views:
 
+    .. versionchanged:: 2.1
+
+    The ``obj`` argument was added.
```
What does the `obj` argument represent at the `ModelAdmin` level? (At the inline level it's the parent... but here?)

jdufresne commented:
> I'm not understanding at all the role of the obj parameter for ModelAdmin. To me these look like distinct methods (with different semantics) that just happen to have the same signature.

Originally I thought it would be desirable to use a single method signature instead. But I see your point that this feature only affects `InlineModelAdmin`. I've removed the use of `obj` from `ModelAdmin`.

jdufresne commented:
Thanks for the review.

I believe I have addressed all feedback. Here is a list of changes:

- Reverted `ModelAdmin.has_add_permission()` changes. The PR now only affects `InlineModelAdmin`
- Reworded to better explain the "If `obj` is `None`" case
- Add a new commit to document `InlineModelAdmin.has_change_permision()` and `InlineModelAdmin.has_delete_permision()`.

All additional feedback is welcome.

carltongibson commented:
Hi @jdufresne. Thanks for updating this. Looks great. (I think the docs additions for `InlineModelAdmin` are a win.)

timgraham commented:
Jon, please check my edits.

jdufresne commented:
LGTM. Thanks Tim.
